app/uk/gov/hmrc/agentsubscriptionfrontend/auth/AuthActions.scala

```diff
@@ -16,83 +16,88 @@
-                case _ =>
-                  Future successful redirectToNonAgentNextSteps
-              }
-            }
-          }
-    }
-
-  def AuthorisedWithAgentAffinityAsync(body: AuthContext => Request[AnyContent] => Future[Result])
-                                         (implicit appConfig: AppConfig): Action[AnyContent] =
-    AuthorisedFor(NoOpRegime, pageVisibility = GGConfidence).async {
-      implicit authContext =>
-        implicit request =>
-          isAgentAffinityGroup.flatMap {
-            case true => body(authContext)(request)
-            case false => Future successful redirectToNonAgentNextSteps
-          }
-    }
-
-  protected def enrolments(implicit authContext: AuthContext, hc: HeaderCarrier): Future[List[Enrolment]] =
-    authConnector.getEnrolments[List[Enrolment]](authContext)
-
-  private def checkActivatedEnrollment(enrolls: List[Enrolment]): Future[Boolean] =
-    enrolls.find(_.key equals "HMRC-AS-AGENT") match {
-      case Some(enroll) if enroll.isActivated => Future successful true
-      case _ => Future successful false
-    }
-
-  protected def isAgentAffinityGroup()(implicit authContext: AuthContext, hc: HeaderCarrier): Future[Boolean] =
-    authConnector.getUserDetails(authContext).map { userDetailsResponse =>
-      val affinityGroup = (userDetailsResponse.json \ "affinityGroup").as[String]
-      affinityGroup == "Agent"
-    }
-
-  private def redirectToNonAgentNextSteps: Result = {
-    mark("Count-Subscription-NonAgent")
-    Redirect(routes.StartController.showNonAgentNextSteps())
+import uk.gov.hmrc.play.bootstrap.config.AuthRedirects
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+class Agent(private val enrolments: Set[Enrolment], private val creds: Credentials) {
+
+  def hasIRPAYEAGENT: Option[Enrolment] = enrolments.find(e => e.key == "IR-PAYE-AGENT" && e.isActivated)
+  def hasIRSAAGENT: Option[Enrolment] = enrolments.find(e => e.key == "IR-SA-AGENT" && e.isActivated)
+
+  def authProviderId = creds.providerId
+  def authProviderType = creds.providerType
+}
+
+object Agent {
+
+  object hasHmrcAsAgentEnrolment {
+    def unapply(agent: Agent): Option[Unit] =
+      if (agent.enrolments.exists(_.key == "HMRC-AS-AGENT")) Some(()) else None
+  }
+
+  object hasNonEmptyEnrolments {
+    def unapply(agent: Agent): Option[Unit] =
+      if (agent.enrolments.nonEmpty) Some(()) else None
   }
 }
+
+trait AuthActions extends AuthorisedFunctions with AuthRedirects with Monitoring {
+
+  def continueUrlActions: ContinueUrlActions
+  def appConfig: AppConfig
+
+  def env = appConfig.environment
+  def config = appConfig.configuration
+
+  def withSubscribingAgent[A](body: Agent => Future[Result])(implicit request: Request[A], hc: HeaderCarrier, ec: ExecutionContext): Future[Result] =
+    authorised(AuthProviders(GovernmentGateway))
+      .retrieve(allEnrolments and affinityGroup and credentials) {
```

Comment:
Small potential to use the `affinityGroup` predicate here to get `client-auth` to check for an agent affinity, rather than a retrieval and manual check via the private `isAnAgent` function.